scope.go

```diff
@@ -261,118 +393,206 @@ func (s *scope) reportLoopRun() {
 
 // reports current registry with scope status lock held
 func (s *scope) reportRegistryWithLock() {
-	s.registry.RLock()
+	s.registry.Lock()
 	if s.reporter != nil {
-		for _, ss := range s.registry.subscopes {
-			ss.report(s.reporter)
+		for name, ss := range s.registry.subscopes {
+			if expired := ss.report(s.reporter); expired {
+				delete(s.registry.subscopes, name)
+				atomic.StoreUint32(&ss.expired, 1)
+			}
 		}
 	} else if s.cachedReporter != nil {
-		for _, ss := range s.registry.subscopes {
-			ss.cachedReport(s.cachedReporter)
+		for name, ss := range s.registry.subscopes {
+			if expired := ss.cachedReport(s.cachedReporter); expired {
+				delete(s.registry.subscopes, name)
+				atomic.StoreUint32(&ss.expired, 1)
+			}
 		}
 	}
-	s.registry.RUnlock()
+	s.registry.Unlock()
+}
+
+// getExpirySeconds gets the expiry period of metrics for this scope
+func (s *scope) getExpirySeconds() int64 {
+	return s.registry.expirePeriodSeconds
+}
+
+// hasExpired indicates whether the scope is expired
+func (s *scope) hasExpired() bool {
+	return atomic.LoadUint32(&s.expired) == 1
+}
+
+// getCurrentScope gets the current active scope which could
+// be this scope or another if another was created after this
+// one expired
+func (s *scope) getCurrentScope() *scope {
+	if s.hasExpired() {
```

Comment:
could the counters/gauges/etc have the same logic, check if they've expired, and if so, get a new counter, and store a pointer to that counter? that could avoid storing `[]*counter` 

not sure if there's a huge benefit either way, but might simplify the code if a single name doesn't have multiple counters

Reply:
Yeah that'll be ideal, but unfortunately it can't be avoided. One can hold a ref to a counter and have it expire (and removed from the current/active scope) and then someone can create a new counter with the same name/tags combination. If they increment both, we now have two references to two counters which both need to be reported, so they have to be stored in a slice of counters.

Reply:
I think you can protect against that scenario with a recursive appproach. Instead of adding to a slice, you could make the current counter hold a reference to the "new" counter,

E.g.,
```go
if c.expired() {
  if newRef := c.scope.hasNew(...); newRef != nil {
    // essentially build a "linked list"
    c.newRef = newRef
  } else {
    // insert back into the scope
    c.scope.counters[...] = c
  }
}

if c.newRef != nil {
  // if we're worried about the linked list getting longer, we could also check whether `newRef` also has
  // a new ref replace `c.newRef` with `newRef.newRef`
  return c.newRef.Inc(...)
}

// actually increment this counter
```

the reporting, expiring in a scope etc all becomes a lot simpler since you won't have a list. we're building a linked list instead, but the maintenance could be simpler.

Reply:
Hmm, it would make the expiry logic a little simpler, but we'll have to always traverse this linkedList to increment anything and since this is an edge case, we'd have to take a hit of checking whether this holds a ref to newRef on every Inc call, which seems like a waste for an extreme edge case (more than one element only exists when they have an old ref and inc it after the expiry period).